src/jison_parser/index.jison

```diff
@@ -28,14 +25,31 @@ expressions
         { return $1; }
     ;
 
+NEWLINE
+    : LINEFEED
+    | SPACE LINEFEED
+    ;
+
+NUM_IDENTIFIER
+    : IDENTIFIER
+    | INTEGER IDENTIFIER { $$ = "" + $INTEGER + $IDENTIFIER }
+    ;
+
+MOVE_IDENTIFIER
+    : NUM_IDENTIFIER
+       { if (!/^[0-9]?[1-9]*[A-Za-z][A-Za-z_]*$/.test($NUM_IDENTIFIER))
```

Comment:
Hmm, if this isn't going to be handled by the grammar, I'd prefer to use a similar validation framework to alg.js:

https://github.com/cubing/alg.js/blob/master/src/validation.ts

Reply:
I'd want this either in a separate class in JS, or handled in some way by the grammar.

Options I see are:
- Split up the lexer part into numbers, moves, etc. and combine them in the grammar composition rules. (similar to `alg.js`)
- Allow very generous move identifiers (e.g. `[0-9A-Za-z_-]+`) and have a validation class in this project.

I think I favor the latter now, since I really don't care to extend the text KSolve parsing format, and would rather make the KPuzzle JSON format aware of slice ranges.

Reply:
Stateful lexers are painful and error-prone and inflexible, so we don't want it in the lexer.

We can put it in the grammar rules, but that's just a nasty merging of concerns (people may want to use kpuzzle without alg or with different algs).  All we'll be doing is having grammar rules to check move syntax, but the move syntax is an alg thing, not a kpuzzle thing.

Easiest would be just to have alg code to validate move syntax.  This is where such validation code belongs.  (One could even argue that alg *wants* this information for the various puzzles; this is probably the best way to handle outliers like y2 for clock).

Probably we simply want every identifier in kpuzzle to be extremely liberal (including underscores, hyphens, numbers, even leading numbers)---I think ksolve and tws are both so liberal.  This introduces a parse conflict between identifiers and numbers, however, that needs to be handled some way.  Using flex-style parsing (longest match, rather than first match) will resolve this.

An eventual extension to kpuzzle may introduce MoveAlg---moves defined from other moves, in which case we may end up doing actual alg-style move parsing of parts of the file.  Not sure how we will handle this.